### PR TITLE
SOV-543: recent trades on graphql

### PR DESCRIPTION
**Pull request by creed-victor**
https://sovryn.atlassian.net/browse/SOV-543

### CODE REVIEW

**Review by soulBit — APPROVED** [review]

**Review by soulBit — DISMISSED** [review]
@creed-victor I noticed that the recent trades query wasn't re-executing, so I've added poll interval param to the relevant hooks

File: src/app/components/RecentTrades/RecentTradeRow.tsx [hunk 1]
```diff
@@ -31,30 +30,26 @@ export const RecentTradeRow: React.FC<RecentTradeRowProps> = ({
     [isOddRow],
   );
 
-  const isLong = useMemo(() => row.loanToken === quoteTokenAddress, [
-    row.loanToken,
-    quoteTokenAddress,
-  ]);
+  const isLong = useMemo(
+    () => row.loanToken.id.toLowerCase() === quoteTokenAddress.toLowerCase(),
```
Comment (Rickk137) [inline src/app/components/RecentTrades/RecentTradeRow.tsx:34]: might be good to have a utility function for this comparison:
```
const compareAddress = (add1: string | null = "", add2: string | null = "") =>
  !!add1 && !!add2 && add1.toLowerCase() === add2.toLowerCase();
 ```

File: src/app/components/RecentTrades/RecentTradeRow.tsx [hunk 2]
```diff
@@ -31,30 +30,26 @@ export const RecentTradeRow: React.FC<RecentTradeRowProps> = ({
...
+    [row.loanToken, quoteTokenAddress],
+  );
 
   const entryPrice = useMemo(() => {
     return isLong
-      ? bignumber(1)
-          .div(row.entryPrice)
-          .mul(10 ** 36)
+      ? bignumber(1).div(row.entryPrice).toString()
```
Comment (Rickk137) [inline src/app/components/RecentTrades/RecentTradeRow.tsx:40]: if for some reason ``row.entryPrice`` is zero then ``entryPrice`` would be infinity

**Review by Rickk137 — APPROVED** [review]

File: src/app/components/RecentTrades/RecentTradeRow.tsx (hunk 2 above)
Comment (creed-victor) [reply src/app/components/RecentTrades/RecentTradeRow.tsx:40]: Hm, I dont think it's possible, but any suggestions on handling infinities? @DistributedCollective/circle-of-light

File: src/app/components/RecentTrades/RecentSwapRow.tsx [hunk 3]
```diff
@@ -22,11 +22,30 @@ export const RecentSwapRow: React.FC<RecentSwapRowProps> = ({
     () => (isOddRow ? 'tw-bg-gray-3' : 'tw-bg-gray-1'),
     [isOddRow],
   );
-  const isBuy = useMemo(() => row.type === SwapsType.BUY, [row.type]);
+  const isBuy = useMemo(
+    () =>
+      row._fromToken.id.toLowerCase() !==
+      baseAssetDetails.getTokenContractAddress().toLowerCase(),
```
Comment (Rickk137) [inline src/app/components/RecentTrades/RecentSwapRow.tsx:28]: Use ``compareAddress`` utility function

**Review by Rickk137 — CHANGES_REQUESTED** [review]

**Comment by Rickk137** [comment]
@creed-victor the build is failing, could you check that as well

**Review by Rickk137 — DISMISSED** [review]

File: src/app/components/RecentTrades/RecentSwapRow.tsx [hunk 4]
```diff
@@ -22,11 +23,32 @@ export const RecentSwapRow: React.FC<RecentSwapRowProps> = ({
...
+    row._return,
+  ]);
+
+  const price = useMemo(
+    () =>
+      isBuy
+        ? Number(row._amount) / Number(row._return)
+        : Number(row._return) / Number(row._amount),
+    [isBuy, row._amount, row._return],
+  );
```
Comment (tiltom) [inline src/app/components/RecentTrades/RecentSwapRow.tsx:47]: The logic here is not correct - I paid `100 XUSD` and received `0,00476816 BPRO`.
image

This is what I see on `testnet` (and it's correct):
image

And this is what I see on `local`: 
image

File: src/app/components/RecentTrades/RecentSwapRow.tsx [hunk 5]
```diff
@@ -38,12 +60,8 @@ export const RecentSwapRow: React.FC<RecentSwapRowProps> = ({
...
         )}
       >
-        <Tooltip
-          position="top"
-          interactionKind="hover"
-          content={<>{row.price}</>}
-        >
-          {Number(row.price).toFixed(quoteAssetDetails.displayDecimals)}
+        <Tooltip position="top" interactionKind="hover" content={<>{price}</>}>
+          {Number(price).toFixed(quoteAssetDetails.displayDecimals)}
```
Comment (tiltom) [inline src/app/components/RecentTrades/RecentSwapRow.tsx:64]: Use `AssetValue` instead with `useTooltip` property.

File: src/app/components/RecentTrades/RecentSwapRow.tsx [hunk 6]
```diff
@@ -52,12 +70,8 @@ export const RecentSwapRow: React.FC<RecentSwapRowProps> = ({
...
         )}
       >
-        <Tooltip
-          position="top"
-          interactionKind="hover"
-          content={<>{row.quoteVolume}</>}
-        >
-          {Number(row.quoteVolume).toFixed(baseAssetDetails.displayDecimals)}
+        <Tooltip position="top" interactionKind="hover" content={<>{size}</>}>
+          {Number(size).toFixed(baseAssetDetails.displayDecimals)}
```
Comment (tiltom) [inline src/app/components/RecentTrades/RecentSwapRow.tsx:74]: Use `AssetValue` instead with `useTooltip` property.

File: src/app/components/RecentTrades/RecentTradeRow.tsx [hunk 7]
```diff
@@ -69,9 +65,9 @@ export const RecentTradeRow: React.FC<RecentTradeRowProps> = ({
         <Tooltip
           position="top"
           interactionKind="hover"
-          content={<>{weiTo18(entryPrice)}</>}
+          content={<>{toNumberFormat(entryPrice, 18)}</>}
         >
-          {weiToAssetNumberFormat(entryPrice, quoteToken)}
+          {toAssetNumberFormat(entryPrice, quoteToken)}
         </Tooltip>
```
Comment (tiltom) [inline src/app/components/RecentTrades/RecentTradeRow.tsx:71]: Use `AssetValue` instead with `useTooltip` property.

File: src/app/components/RecentTrades/RecentTradeRow.tsx [hunk 8]
```diff
@@ -83,9 +79,9 @@ export const RecentTradeRow: React.FC<RecentTradeRowProps> = ({
         <Tooltip
           position="top"
           interactionKind="hover"
-          content={<>{weiTo18(positionSize)}</>}
+          content={<>{toNumberFormat(positionSize, 18)}</>}
         >
-          {weiToAssetNumberFormat(positionSize, baseToken)}
+          {toAssetNumberFormat(positionSize, baseToken)}
         </Tooltip>
```
Comment (tiltom) [inline src/app/components/RecentTrades/RecentTradeRow.tsx:85]: Use `AssetValue` instead with `useTooltip` property.

File: src/app/hooks/trading/useSwap_RecentTrades.ts [hunk 9]
```diff
@@ -1,36 +1,32 @@
...
+    () => LiquidityPoolDictionary.get(baseToken, quoteToken),
+    [baseToken, quoteToken],
   );
-  const baseTokenAddress = getTokenContract(baseToken).address;
-  const quoteTokenAddress = getTokenContract(quoteToken).address;
-  const [loading, setLoading] = useState(false);
+
+  const [load, { called, data, loading }] = useGetRecentSwapEventsLazyQuery({
+    variables: {
+      converterAddress: [pool.converter].map(item => item.toLowerCase()),
```
Comment (tiltom) [inline src/app/hooks/trading/useSwap_RecentTrades.ts:16]:
```suggestion
      converterAddress: [pool.converter.toLowerCase()],
```
Why you did it via `map` before?

File: src/utils/helpers.ts [hunk 10]
```diff
@@ -81,6 +81,9 @@ export const isAddress = (address: string) => {
   }
 };
 
+export const compareAddress = (address1: string, address2: string) =>
```
Comment (tiltom) [inline src/utils/helpers.ts:84]:
```suggestion
export const areAddressesEqual = (address1: string, address2: string) =>
```

**Review by tiltom — CHANGES_REQUESTED** [review]

File: src/app/hooks/trading/useSwap_RecentTrades.ts (hunk 9 above)
Comment (creed-victor) [reply src/app/hooks/trading/useSwap_RecentTrades.ts:16]: Ops, there was multiple addresses before :)

**Review by tiltom — CHANGES_REQUESTED** [review]
I see no recent trades on `Spot` (`Margin` works fine), @creed-victor pls fix it first.

**Comment by creed-victor** [comment]
> I see no recent trades on `Spot` (`Margin` works fine), @creed-victor pls fix it first.

@tiltom, done

**Review by tiltom — CHANGES_REQUESTED** [review]
@creed-victor This is what I see on testnet (and should see), the first row contains my trade:
image

And this is what I see on local (my last order isn't there, even after refreshing multiple times):
image

Pls, make sure that those 2 envs are identical before you submit it for re-review.

**Comment by creed-victor** [comment]
> @creed-victor This is what I see on testnet (and should see), the first row contains my trade: image
> 
> And this is what I see on local (my last order isn't there, even after refreshing multiple times): image
> 
> Pls, make sure that those 2 envs are identical before you submit it for re-review.

I think endpoint just doesn't have data on bnb/rbtc, other pairs works as they should

**Review by tiltom — DISMISSED** [review]

**Review by soulBit — APPROVED** [review]

**Review by LunaSov — DISMISSED** [review]

**Pull request merged**